webapp/src/http_handlers/sessions/session.rs

```diff
@@ -0,0 +1,161 @@
+use chrono::{DateTime, Utc};
+use rocket::{Request, http::{Cookie, CookieJar, Status},
+    request::{self, FromRequest}};
+use serde::{Deserialize, Serialize};
+use std::{hash::{Hash, Hasher}, path::{Path, PathBuf}, sync::{Arc, Mutex}};
+
+use crate::{config_struct::BackendConfig, filework::new_session_folder};
+use super::sessions_tracker::SessionsTracker;
+
+// Anonymous session data guard
+#[derive(Eq, Clone, Serialize, Deserialize)]
+pub struct Session
+{
+    pub id: u128,
+    pub last_connection: DateTime<Utc>,
+    pub folder: PathBuf
+}
+
+impl Session
+{
+    pub fn with_id(id: u128) -> Session
+    {
+        Session
+        {
+            id,
+            last_connection: Utc::now(),
+            folder: PathBuf::new()
+        }
+    }
+
+    pub fn folder(mut self, parent_folder: &Path) -> Session
+    {
+        self.folder = parent_folder.to_owned();
+        self
+    }
+
+    pub(crate) fn establish_new(
+        cookies: &CookieJar<'_>, 
+        tracker: &mut SessionsTracker,
+        parent_folder: &Path,
+        logger: &slog::Logger) 
+        -> Option<u128>
+    {
+        let session_id = uuid::Uuid::new_v4().as_u128();
+        let session_id_str = session_id.to_string();
+        let folder;
+        match new_session_folder(parent_folder, &session_id_str,
+            logger)
+        {
+            Some(new_folder) => 
+            {
+                folder = new_folder;
+            },
+            None => 
+            {
+                return None;
+            }
+        }
+        
+        tracker.insert_session(Session::with_id(session_id)
+            .folder(&folder));
+        cookies.add(Cookie::new("session_id", session_id_str));
+        info!(logger, "New session established: {}", session_id);
+
+        Some(session_id)
+    }
+}
+
+#[rocket::async_trait]
+impl<'r> FromRequest<'r> for Session
+{
+    type Error = ();
+
+    async fn from_request(req: &'r Request<'_>) -> request::Outcome<Self, Self::Error>
+    {
+        let logger = req.rocket().state::<slog::Logger>().unwrap();
+        // TODO: there theoretically is a case when mutex locking fails
```

Comment:
as far as I remember, Mutex can't be locked if it was poisoned (when sibling thread who acquired the Mutex crashed)

Reply:
Yes, this is what I meant here, but I left it as a TODO for now because there are no obviously dangerous usages of this mutex in the code. And also I don't know what to do in such cases yet, sessions tracker fail is fatal for the server and I haven't thought of what to do when it happens.
I would be grateful for suggestions here.